keps/sig-network/2091-admin-network-policy/README.md

```diff
@@ -0,0 +1,1116 @@
+with ipBlock rules that select external traffic.
+
+## Proposal
+
+In order to achieve the three primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `policy.networking.k8s.io` API group:
+- AdminNetworkPolicy
+
+### AdminNetworkPolicy resource
+
+An AdminNetworkPolicy (ANP) resource will help the administrators:
+1. Set strict security rules for the cluster, i.e. a developer CANNOT override
+these rules by creating NetworkPolicies that applies to the same workloads as the
+AdminNetworkPolicy does.
+2. Set baseline security rules that describes default connectivity for cluster
+workloads, which CAN be overridden by developer NetworkPolicies if needed.
+
+#### Actions
+
+Unlike the NetworkPolicy resource in which each rule represents an allowed
+traffic, AdminNetworkPolicy will enable administrators to set `Pass`,
+`Deny` or `Allow` as the action of each rule. AdminNetworkPolicy rules should
+be read as-is, i.e. there will not be any implicit isolation effects for the Pods
+selected by the AdminNetworkPolicy, as opposed to what NetworkPolicy rules imply.
+
+- Pass: Traffic that matches a `Pass` rule will skip all further rules from all 
+  ANPs and instead be enforced by the K8s NetworkPolicies. If there is no K8s 
+  NetworkPolicy rule match, traffic will be allowed.
+- Deny: Traffic that matches a `Deny` rule will be dropped. 
+- Allow: Traffic that matches an `Allow` rule will be allowed.
+
+AdminNetworkPolicy `Deny` rules are useful for administrators to explicitly
+block traffic with malicious clients, or workloads that poses security risks.
+Those traffic restrictions can only be lifted once the `Deny` rules are deleted
+or modified by the admin.
+
+On the other hand, the `Allow` rules can be used to call out traffic in the cluster
+that needs to be allowed for certain components to work as expected (egress to
+CoreDNS for example). Those traffic should not be blocked when developers apply
+NetworkPolicy to their Namespaces which isolates the workloads.
+
+AdminNetworkPolicy `Pass` rules allows an admin to delegate security posture for
+certain traffic to the Namespace owners by overriding any lower priority Allow or 
+Deny rules. For example, allowing intra-tenant traffic can be delegated to tenant 
+admins explicitly by the cluster admin with the use of `Pass` rules.
+
+#### Priority
+
+The policy instances will be ordered based on the numeric priority assigned to the
+ANP. A larger number corresponds to a higher precedence. For example, ANP with `priority`
+set to "1" will be considered as the policy with the lowest precedence, while "999" will
+be considered as highest precedence. Any positive priority, however, will have
+higher precedence over the namespaced NetworkPolicy instances in the cluster.
+
+Additionally, the special priority "0" can be used in the priority field to indicate
```

Comment:
I meeeean you _could_ look at it that way in the order of things.... 

> I can't yet think of anything better, but my gut tells me an implicit value of 0 keying a substantial behavioral difference is going to be bad.

We could get rid of the overloaded "0" priority and replace it with a boolean however the only use case we forsee involving the "baseline" priority "0" is flipping the default cluster stance from allow to deny

```
#### Story 5: Cluster Wide Default Guardrails

As a cluster admin I want to change the default security model for my cluster, 
so that all intra-cluster traffic (except for certain essential traffic) is 
blocked by default. Namespace owners will need to use NetworkPolicies to 
explicitly allow known traffic. This follows a whitelist model which is 
familiar to many security administrators, and similar
to how [kubernetes suggests network policy be used](
``` 
And this implementation seemed like the easiest way to satisfy that story

Reply:
default is a good point.

Reply:
Another idea: it seems to be two cases that we want to express, w.r.t. the interaction between ANP and NP, beyond "force block" and "force allow":

1: Defer exclusively to NetworkPolicy. IOW, the `Pass` effect.
2: Defer to NetworkPolicy, **but default to deny if there are no NPs written**

What if we add a fourth match action that would express case 2? Then we would have

- `Deny: exclusively deny
- `ForceAllow: always allow
- `Pass: "pass down" to NP
- `Defer`: "pass down" to NP, but "force" there to be a default-deny NP if it doesn't exist.

Alternatively, reserve priority "500" for NetworkPolicy, and don't allow ANPs with this priority.

Reply:
maybe add an UNRESOLVED here

Reply:
Agreed and Done

Reply:
@thockin That's a really good question :) Currently I would say we don't allow a priority=0 rule with a `pass` action but for now I think it's alright